### ISSUE TITLE
Structure and consistency improvements

### ISSUE THREAD

**Issue by networkException** [purecoreio/purecore-js#1]
**Introduction**

The current version of the wrapper has various issues when it comes to the structure and consistency of the code. For once, there is no typesafe differentiation between a Session using a hash as the tool or using a key, which effectively replaces the network or instance parameter on most of not all requests. To deal with those issues I am proposing the following structure:

**Structure**

The Core object still serves as the root of a "chain". The constructing a core object requires the tool. The core object itself only contains methods for accessing endpoints, if no authentication is required for said endpoints. The other methods existing on the core object are "as" methods. 

`asOwner(): OwnerSession` for example returns a typesafe object for owner session related actions. As the type system cannot check if the entered tool is really a hash, this gets assumed in the as method. The owner session in this case would also extend a normal session to allow access to non owner methods, such as requesting the own discord id. 

As a hash is the tool for a session, I would propose that a key is the tool for an application, as such the `asApplication(): Application` method can be called on the core object. 

Further objects returned for example by the session don't have to provide a tool and just context parameters. An `async getPlayer(id: string): Promise<Player>` method for example would allow a session to request a player. The normal player object however will not provide access to a `async getConnections(page: number): Promise<Array<Connection>>`, as this would require a network context. The following examples would provide such a context:

`new Core("key").asApplication().getPlayer("id");`
`new Core("hash").asSession().getNetwork("id").getPlayer("id");`

(example simplified ofc, methods requesting data are async and would need either an await or .then(). Also for some reason as of right now the network is not required for the endpoint even with a hash)

**Consistency**

The following would be my idea for a type class, serving no more purpose than representing an object. (No further request methods)

```ts
class SearchResult {
    //Field are private and readonly, also have types
    private readonly id: string;
    private readonly type: SearchResultType;
    private readonly name: string;

    //The constructors only purpose is to take arguments
    public constructor(id: string, type: SearchResultType, name: string) {
        this.id = id;
        this.type = type;
        this.name = name;
    }

    //Getters like this. This may not resemble the typescript guidelines but it better fits the style other languages have
    public getId(): string {
        return this.id;
    }

    public getType(): SearchResultType {
        return this.type;
    }

    public getName(): string {
        return this.name;
    }

    //A static method to parse json to the type
    public static fromJSON(json: any): SearchResult {
        return new SearchResult(
            json.uuid,
            SearchResultType[json.type as keyof typeof SearchResultType],
            json.name
        );
    }
}

```

A method accessing an endpoint would look like this. 

```ts
//Async method with types
public async search(query: string): Promise<Array<SearchResult>> {
    return CoreRequest.get<Array<SearchResult>>(
        "search", //The endpoint used
        new ParameterStack() //The parameters given, using the parameterstack object
            .push("string", query),
        json => json.forEach(SearchResult.fromJSON) //This parses the json to the required objects
    );
}
```

I already started to implement some of the mentioned changed into my fork of the wrapper, however I'd like to have some feedback and if such a structure could be used in a release of the wrapper

**Comment by networkException** [comment]
Ok I have been working on updating the current wrapper, though I have found many inconsistencies regarding parameters. Lets for example take `https://api.purecore.io/rest/2/player/connection/list/`. The backend returns `key, player, page? | hash, network, player, page?` as the required parameters, though in the wrapper it looks more like `key, player, instance?, page? | hash, player, instance?, page?`. From my understanding connections are instance based and punishments as well as payments network based. This however would mean that said endpoint would require something like this `key, player, page? | hash, instance, player, page?`. 

Not just the wrapper seems to not follow this idea, as requesting `https://api.purecore.io/rest/2/player/connection/list/?hash=<>&page=0&player=<>` returns a list of connections - if they are global or default to an instance I don't know

It would be helpful if there would be clarification on those topics, I generally just would need guidelines to follow what endpoints can be accessed by who and in what ways

**Comment by networkException** [comment]
Partly implemented changes in #2, still lacking private readonly fields due to backwards compatibility. I would propose more changes to be discussed for a v2 release on npm

**Comment by networkException** [comment]
Similar to discordjs, I would propose resolvable types for api methods. This way, a user could for example use a player object or a string representing a players coreid, as both would provide a way for the api to identify the player uniquely (the player object would be processed by the wrapper to extract the id anyways). 

This resolvable type would be implemented for any object which has an id or another common way of identification in the api
```typescript
type GatewayResolvable: Gateway | string

class Store {
    public async unlinkGateway(gateway: GatewayResolvable) {
        //...
    }
}

const gateway: Gateway = //...

store.unlinkGateway(gateway); //would be ok (the method would call a getId() method for use as an endpoint argument)
store.unlinkGateway("paypal"); //would be ok (would be directly used as an argument for the endpoint)
```

I would aim for those changes to be implemented after the close if #2, which currently is lacking documentation and refactoring of analytics

**Comment by quiquelhappy** [comment]
https://github.com/purecoreio/purecore-js/pull/2
